pkg/util/retryutil/retry_util.go

```diff
@@ -0,0 +1,40 @@
+package retryutil
+
+import (
+	"fmt"
+	"time"
+)
+
+type RetryError struct {
+	n int
+}
+
+func (e *RetryError) Error() string {
+	return fmt.Sprintf("still failing after %d retries", e.n)
+}
+
+type ConditionFunc func() (bool, error)
+
+// Retry retries f every interval until after maxRetries.
+// The interval won't be affected by how long f takes.
+// However, if f takes longer than interval, it will be delayed.
+// For example, if interval is 3s, f takes 1s, another f will be called 2s later.
+func Retry(interval time.Duration, maxRetries int, f ConditionFunc) error {
+	tick := time.NewTicker(interval)
+	defer tick.Stop()
+
+	for i := 0; i < maxRetries; i++ {
+		ok, err := f()
+		if err != nil {
+			return err
+		}
+		if ok {
+			return nil
+		}
+		if i+1 == maxRetries {
```

Comment:
we should either get rid of condition checking (i < maxRetries) at line 26 or this